otter/test/convergence/test_service.py

```diff
@@ -1,95 +1,409 @@
+            return Effect(Constant('foo'))
 
-        def execute_convergence(group, desired, lc, now, log):
-            return Effect(Constant((group, desired, lc, now, log)))
+        my_buckets = [0, 5]
+        converger = Converger(
+            self.log, self.dispatcher, self.buckets,
+            self._pfactory, converge_all=converge_all)
+
+        result = self.fake_partitioner.got_buckets(my_buckets)
+        self.assertEqual(self.successResultOf(result), 'foo')
+        self.log.msg.assert_called_once_with(
+            'buckets-acquired', my_buckets=[0, 5], system='converger')
+
+    def test_buckets_acquired_errors(self):
+        """
+        Errors raised from performing the converge_all effect are logged, and
+        None is the ultimate result.
+        """
+        def converge_all(log, group_locks, _my_buckets, all_buckets):
+            return Effect(Error(RuntimeError('foo')))
+
+        Converger(
+            self.log, self.dispatcher, self.buckets,
+            self._pfactory, converge_all=converge_all)
+
+        result = self.fake_partitioner.got_buckets([0])
+        self.assertEqual(self.successResultOf(result), None)
+        self.log.err.assert_called_once_with(
+            CheckFailureValue(RuntimeError('foo')),
+            'converge-all-error', system='converger')
+
+
+class ConvergeOneTests(SynchronousTestCase):
+
+    def setUp(self):
+        self.log = mock_log()
+        self.tenant_id = 'tenant-id'
+        self.group_id = 'g1'
+        self.version = 5
+        self.deletions = []
+        self.dispatcher = ComposedDispatcher([
+            EQFDispatcher([
+                (DeleteNode(path='/groups/divergent/tenant-id_g1',
+                            version=self.version),
+                 lambda i: self.deletions.append(True))
+            ]),
+            _get_dispatcher(),
+        ])
+
+    def test_success(self):
+        """
+        runs execute_convergence and returns None, then deletes the dirty flag.
+        """
+        calls = []
+
+        def execute_convergence(tenant_id, group_id, log):
+            return Effect(Func(
+                lambda: calls.append((tenant_id, group_id, log))))
+
+        eff = converge_one(
+            self.log, make_lock_set(), self.tenant_id, self.group_id,
+            self.version,
+            execute_convergence=execute_convergence)
+        result = sync_perform(self.dispatcher, eff)
+        self.assertEqual(result, None)
+        self.assertEqual(
+            calls,
+            [(self.tenant_id, self.group_id,
+              matches(IsBoundWith(tenant_id=self.tenant_id,
+                                  group_id=self.group_id)))])
+        self.assertEqual(self.deletions, [True])
+        self.log.msg.assert_any_call(
+            'mark-clean-success',
+            tenant_id=self.tenant_id, group_id=self.group_id)
+
+    def test_non_concurrent(self):
+        """Runs execute_convergence non-concurrently, based on the group ID."""
```

Comment:
Apologies, this was kind of confusing to me - maybe just "Do not run execute_convergence again, if the group is already converging."?  Since this test isn't actually testing that convergence will be run again later.